File: supabase/functions/scrape-event/sources/axs.ts
import { assert } from "../../_shared/http.ts";
import { EventVisibility } from "../../_shared/squadquest.ts";
import { Event } from "../../_shared/squadquest.ts";

function canScrape(url: URL): boolean {
  return (
    url.hostname === "www.axs.com" &&
    url.pathname.toLowerCase().startsWith("/events/")
  );
}

async function scrape(url: URL): Promise<Event> {
  // Import puppeteer
  const { default: puppeteer } = await import(
    "https://deno.land/x/puppeteer@16.2.0/mod.ts"
  );

  // Launch browser
  const browser = await puppeteer.connect({
    browserWSEndpoint: "ws://chrome:3000",
  });

  try {
    // Create a new page with stealth settings
    const page = await browser.newPage();

    // Navigate to URL
    await page.goto(url.toString(), {
      waitUntil: "domcontentloaded",
      timeout: 60000,
    });

    // Get page content
    const html = await page.content();

    // Find all ld+json scripts and look for the one containing event data
    const scripts =
      html.match(/<script type="application\/ld\+json">(.*?)<\/script>/gs) ||
      [];
    let eventData = null;

    for (const script of scripts) {
      const jsonContent = script.replace(
        /<script type="application\/ld\+json">/,
        "",
      ).replace(/<\/script>/, "");
      try {
        const data = JSON.parse(jsonContent);
        if (data["@type"] === "Event" || data["@type"] === "MusicEvent") {
          eventData = data;
          break;
        }
      } catch (e) {
        continue;
      }
    }

    assert(eventData != null, "Failed to extract event data from AXS");

    try {
      // build event object
      const startTime = eventData.startDate
        ? new Date(eventData.startDate)
        : undefined;
      assert(startTime != null, "Event start time is required");

      return {
        title: eventData.name || "Untitled Event",
        start_time_min: startTime,
        start_time_max: startTime
          ? new Date(startTime.getTime() + 15 * 60 * 1000)
          : undefined,
        end_time: eventData.endDate ? new Date(eventData.endDate) : undefined,
        location_description: eventData.location?.name ||
          eventData.location?.address?.addressLocality,
        link: url.toString(),
        notes: eventData.description || undefined,
        banner_photo: typeof eventData.image === "string"
          ? eventData.image.startsWith("//")
            ? `https:${eventData.image}`
            : eventData.image
          : undefined,
        visibility: EventVisibility.public,
      };
    } catch (error) {
      throw new Error(`Failed to parse AXS event data: ${error.message}`);
    }
  } finally {
    await browser.close();
  }
}

export default { canScrape, scrape };

File: lib/services/scrapers/axs.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:squadquest/models/instance.dart';

class AxsScraper {
  static bool canHandle(Uri url) {
    return (url.host == 'www.axs.com' || url.host == 'axs.com') &&
        url.path.startsWith('/events/');
  }

  static Future<Instance> scrape(Uri url) async {
    final response = await http.get(url, headers: {
      'Accept':
          'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
      'User-Agent':
          'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
    });

    if (response.statusCode != 200) {
      throw Exception('Failed to load event page from AXS');
    }

    // Extract JSON-LD data
    final regex = RegExp(
        r'<script type="application\/ld\+json">(.*?)<\/script>',
        multiLine: true,
        dotAll: true);
    final match = regex.firstMatch(response.body);

    if (match == null) {
      throw Exception('Failed to find event data on page');
    }

    final jsonStr = match.group(1)!;
    final eventData = json.decode(jsonStr);

    if (eventData['@type'] != 'MusicEvent' && eventData['@type'] != 'Event') {
      throw Exception('Not a valid event page');
    }

    final startTime = DateTime.parse(eventData['startDate']);

    return Instance(
      title: eventData['name'] ?? 'Untitled Event',
      startTimeMin: startTime,
      startTimeMax: startTime.add(const Duration(minutes: 15)),
      endTime: eventData['endDate'] != null
          ? DateTime.parse(eventData['endDate'])
          : null,
      locationDescription: eventData['location']?['name'] ??
          eventData['location']?['address']?['addressLocality'] ??
          '',
      link: url,
      notes: eventData['description'],
      bannerPhoto: eventData['image'] != null
          ? Uri.parse(eventData['image'].toString().startsWith('//')
              ? 'https:${eventData['image']}'
              : eventData['image'])
          : null,
      visibility: InstanceVisibility.public,
    );
  }
}
